File: heap-vs-stack/heap_example.cpp
#include <iostream>
#include <memory>

void heapAllocation() {
    std::cout << "=== Heap Allocation Example ===" << std::endl;
    std::unique_ptr<int[]> largeArray(new int[1000000]);
    largeArray[0] = 42;
    largeArray[999999] = 100;
    std::cout << "Large heap allocation succeeded!" << std::endl;
    std::cout << "First element: " << largeArray[0] << std::endl;
    std::cout << "Last element: " << largeArray[999999] << std::endl;
}

void deepRecursionWithHeap(int depth, int maxDepth) {
    auto data = std::make_unique<int>(depth);
    if (depth % 1000 == 0) {
        std::cout << "Recursion depth: " << depth << std::endl;
    }
    if (depth < maxDepth) {
        deepRecursionWithHeap(depth + 1, maxDepth);
    }
}

int main() {
    heapAllocation();
    std::cout << "\n=== Deep Recursion with Heap Storage ===" << std::endl;
    deepRecursionWithHeap(0, 10000);
    std::cout << "Completed 10,000 recursive calls successfully!" << std::endl;
    return 0;
}

File: heap-vs-stack/heap_example.rs
fn heap_allocation() {
    println!("=== Heap Allocation Example ===");
    let large_array = Box::new([0i32; 1_000_000]);
    println!("Large heap allocation succeeded!");
    println!("First element: {}", large_array[0]);
    println!("Last element: {}", large_array[999_999]);
}

fn deep_recursion_with_heap(depth: u64, max_depth: u64) {
    let data = Box::new(depth);
    if depth % 1000 == 0 {
        println!("Recursion depth: {}", depth);
    }
    if depth < max_depth {
        deep_recursion_with_heap(depth + 1, max_depth);
    }
}

fn dynamic_heap_allocation() {
    println!("\n=== Dynamic Heap Allocation with Vec ===");
    let mut vec = Vec::new();
    for i in 0..1_000_000 {
        vec.push(i);
    }
    println!("Vec with {} elements allocated on heap", vec.len());
    println!("First element: {}", vec[0]);
    println!("Last element: {}", vec[vec.len() - 1]);
}

fn main() {
    heap_allocation();
    println!("\n=== Deep Recursion with Heap Storage ===");
    deep_recursion_with_heap(0, 10_000);
    println!("Completed 10,000 recursive calls successfully!");
    dynamic_heap_allocation();
}
